src/pobox.erl

```diff
@@ -60,35 +105,70 @@
 %% up based on guards.
--spec start_link(name(), max(), 'stack' | 'queue' | keep_old | {mod, module()}, 'notify'|'passive') -> {ok, pid()}
+-spec start_link(name(), max(), stack | queue | keep_old | {mod, module()}, notify | passive) -> {ok, pid()}
         ;       (term(), pid(), max(), stack | queue) -> {ok, pid()}.
-start_link(Owner, Size, Type, StateName) when is_pid(Owner);
-                                              is_atom(Owner),
+start_link(Owner, Size, Type, StateName) when ?PROCESS_NAME_GUARD(Owner),
+                                              ?POBOX_START_STATE_GUARD(StateName),
                                               is_integer(Size), Size > 0 ->
-    gen_statem:start_link(?MODULE, {Owner, Size, Type, StateName}, []);
-start_link(Name, Owner, Size, Type) ->
+    gen_statem:start_link(?MODULE, #pobox_opts{owner = Owner, size = Size, type = Type, state_name = StateName}, []);
+start_link(Name, Owner, Size, Type)
+  when Size > 0,
+    ?PROCESS_NAME_GUARD_WITH_LOCAL_NO_PID(Name),
+    ?PROCESS_NAME_GUARD(Owner),
+    ?POBOX_BUFFER_TYPE_GUARD(Type) ->
     start_link(Name, Owner, Size, Type, notify).
 
--spec start_link(name(), pid(), max(), stack | queue | keep_old | {mod, module()},
+-spec start_link(name(), name(), max(), stack | queue | keep_old | {mod, module()},
                  'notify'|'passive') -> {ok, pid()}.
-start_link(Name, Owner, Size, Type = {mod, Module}, StateName) when Size > 0,
-                                                                is_atom(Module),
-                                                                StateName =:= notify orelse
-                                                                StateName =:= passive ->
-    gen_statem:start_link(Name, ?MODULE, {Owner, Size, Type, StateName}, []);
-start_link(Name, Owner, Size, Type, StateName) when Size > 0,
-                                                    Type =:= queue orelse
-                                                    Type =:= stack orelse
-                                                    Type =:= keep_old,
-                                                    StateName =:= notify orelse
-                                                    StateName =:= passive ->
-    gen_statem:start_link(Name, ?MODULE, {Owner, Size, Type, StateName}, []).
+start_link(Name, Owner, Size, Type, StateName)
+  when Size > 0,
+    ?PROCESS_NAME_GUARD_WITH_LOCAL_NO_PID(Name),
+    ?PROCESS_NAME_GUARD(Owner),
+    ?POBOX_BUFFER_TYPE_GUARD(Type),
+    ?POBOX_START_STATE_GUARD(StateName) ->
+    gen_statem:start_link(Name, ?MODULE, #pobox_opts{
+        name = Name,
+        owner = Owner,
+        size = Size,
+        type = Type,
+        state_name = StateName
+    }, []).
+
+-ifndef(no_map_type). %% If we have map support
+default_opts() ->
+  #{
+    owner => self(),
+    state_name => notify,
+    size => 100,
+    type => queue,
+    heir => none,
+    heir_data => none
+  }.
+
+start_link(Opts) when is_map(Opts) ->
```

Comment:
Provide an equivalent mechanism for Erlang version R16 and older. Speaking of this we might want to use Travis CI to check multiple versions of Erlang simultaneously. What do you think?